File: solutions/0/0/1/4/p0014_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 14: Longest Collatz Sequence [Level 0]. """
from __future__ import annotations

import functools
import sys
from sys import argv, stderr
from time import perf_counter
from typing import Any


@functools.lru_cache(maxsize=None)
def collatz_sequence_length(number: int) -> int:
    """Calculate the Collatz sequence length recursively with memoization."""
    if number == 1:
        return 1
    elif number % 2 == 0:
        return 1 + collatz_sequence_length(number // 2)
    else:
        return 1 + collatz_sequence_length(3 * number + 1)


def plot_collatz_sequence_lengths_upto(number: int) -> None:
    """Plot the Collatz sequence lengths up to a given number."""
    import matplotlib.pyplot as plt

    numbers = range(1, number + 1)
    lengths = [collatz_sequence_length(i) for i in numbers]
    plt.title("Collatz Sequence Lengths")
    plt.plot(numbers, lengths)
    plt.xlabel("Number")
    plt.ylabel("Collatz Sequence Length")
    plt.tight_layout()
    plt.show()


def solve(*, max_number: int) -> int:
    max_length, starting_number = (0, 0)
    power_of_two: int = 2 ** (max_number.bit_length() - 1)
    for x in range(max_number, power_of_two, -1):
        length = collatz_sequence_length(x)
        if length > max_length:
            max_length, starting_number = (length, x)
    if sys.argv[-1] == "--show":
        plot_collatz_sequence_lengths_upto(number=max_number)
    return starting_number


def main(**kwargs: Any) -> int:
    """
    Usage: ./file.py <kwarg>... [--runs=1] [--show]
    Output: "<runs> <avg_seconds> <result>"
    """
    try:
        runs_arg: str = next((arg for arg in argv[1:] if arg.startswith("--runs=")))
        runs: int = int(runs_arg.split("=", 1)[1])
        assert runs > 0
    except (AssertionError, StopIteration, ValueError):
        runs = 1
    elapsed: list[float] = []
    result: int | None = None
    rc: int = 0
    errors: list[str] = []
    for _ in range(runs):
        _start, _result, _stop = (perf_counter(), solve(**kwargs), perf_counter())
        elapsed.append(_stop - _start)
        if result is not None and _result != result:
            errors.append(f"Expected consistent result, got {_result} previous result={result}")
        result = _result
    if result is None:
        errors.append("Expected a result, got None")
    average: float = sum(elapsed) / len(elapsed)
    if errors:
        print("\n".join(errors), file=stderr)
        rc = 1
    print(f"{runs} {average} {result}")
    return rc


if __name__ == "__main__":
    raise SystemExit(main(max_number=int(argv[1])))

File: solutions/0/0/1/4/p0014_s0.c
/* Solution to Euler Problem 14: Longest Collatz Sequence. */
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <time.h>

static long long *cache = NULL;
static long long cache_size = 0;

static long long collatz_length(long long n) {
    if (n == 1) return 1;
    if (n < cache_size && cache[n] != 0) return cache[n];

    long long result;
    if (n % 2 == 0) {
        result = 1 + collatz_length(n / 2);
    } else {
        result = 1 + collatz_length(3 * n + 1);
    }

    if (n < cache_size) {
        cache[n] = result;
    }
    return result;
}

long long solve(int argc, char *argv[]) {
    long long max_number = (argc > 1) ? atoll(argv[1]) : 1000000LL;

    cache_size = max_number + 1;
    cache = (long long *)calloc((size_t)cache_size, sizeof(long long));
    if (!cache) {
        fprintf(stderr, "Out of memory\n");
        return -1;
    }

    /* Find largest power of two strictly below max_number */
    long long power_of_two = 1;
    while (power_of_two * 2 < max_number) {
        power_of_two *= 2;
    }
    /* power_of_two is now the largest power of two < max_number */

    long long max_length = 0, starting_number = 0;
    for (long long x = max_number; x > power_of_two; x--) {
        long long length = collatz_length(x);
        if (length > max_length) {
            max_length = length;
            starting_number = x;
        }
    }

    free(cache);
    cache = NULL;
    cache_size = 0;

    return starting_number;
}

/* Usage: ./file <kwarg>... [--runs=1] [--show]
 * Output: "<runs> <avg_seconds> <result>" */
int main(int argc, char *argv[]) {
    int runs = 1;

    char **solve_argv = malloc((size_t)argc * sizeof(char *));
    if (!solve_argv) {
        fprintf(stderr, "runner: out of memory\n");
        return 1;
    }
    int solve_argc = 0;
    solve_argv[solve_argc++] = argv[0];

    for (int i = 1; i < argc; i++) {
        if (argv[i][0] == '\0') continue;
        if (strncmp(argv[i], "--runs=", 7) == 0) {
            int r = atoi(argv[i] + 7);
            if (r >= 1) runs = r;
            continue;
        }
        if (strcmp(argv[i], "--show") == 0) continue;
        solve_argv[solve_argc++] = argv[i];
    }

    long long result = 0;
    double total = 0.0;
    int rc = 0;
    int has_result = 0;

    for (int r = 0; r < runs; r++) {
        struct timespec t0, t1;
        clock_gettime(CLOCK_MONOTONIC, &t0);
        long long cur = solve(solve_argc, solve_argv);
        clock_gettime(CLOCK_MONOTONIC, &t1);
        total += (double)(t1.tv_sec - t0.tv_sec)
               + (double)(t1.tv_nsec - t0.tv_nsec) * 1e-9;
        if (has_result && cur != result) {
            fprintf(stderr, "Expected consistent result, got %lld previous result=%lld\n",
                    cur, result);
            rc = 1;
        }
        result = cur;
        has_result = 1;
    }

    free(solve_argv);
    printf("%d %.17g %lld\n", runs, total / (double)runs, result);
    return rc;
}